File: hacking_with_python/CSP.py
import requests
from bs4 import BeautifulSoup
import re

# Function to test CSP bypass for a given URL
def test_csp_bypass(url):
    print("\033[95m[*] Testing Content Security Policy (CSP) Bypass...\033[0m")
    
    # Send a GET request to the target URL
    response = requests.get(url)
    print("\033[94m[*] Sending GET request to:", url, "\033[0m")
    
    # Parse the response HTML
    soup = BeautifulSoup(response.content, 'html.parser')
    
    # Check if CSP is present in the response headers
    if 'content-security-policy' in response.headers:
        print("\033[92m[*] Content Security Policy (CSP) found in response headers.\033[0m")
        csp_header = response.headers['content-security-policy']
        print("\033[93m[*] Content Security Policy (CSP) Header:", csp_header, "\033[0m")
        
        # Check if 'unsafe-inline' is present in CSP header
        if "'unsafe-inline'" in csp_header:
            print("\033[91m[!] Target website is potentially vulnerable to CSP bypass.\033[0m")
            
            # Extract URLs with 'unsafe-inline' from HTML content
            inline_scripts = soup.find_all('script', attrs={'src': False})
            for script in inline_scripts:
                if script.text.strip() != "":
                    print("\033[96m[*] Potentially bypassable inline script found:\033[0m", script.text.strip())
            
            # Additional checks or techniques to test CSP bypass can be added here
            
        else:
            print("\033[92m[*] Content Security Policy (CSP) does not contain 'unsafe-inline'.\033[0m")
            print("\033[92m[*] Target website is not vulnerable to CSP bypass.\033[0m")
    else:
        print("\033[91m[!] Content Security Policy (CSP) not found in response headers.\033[0m")
        print("\033[92m[*] Target website may or may not be vulnerable to CSP bypass.\033[0m")

# Main function to execute the program
def main():
    print("\033[96m[>] Content Security Policy (CSP) Bypass Tester\033[0m")
    target_url = input("\033[93m[*] Enter your target website URL: \033[0m")
    test_csp_bypass(target_url)

if __name__ == "__main__":
    main()

File: hacking_with_javascript/CSP.js
const fetch = require('node-fetch');
const colors = require('colors');

// Function to test CSP bypass
async function testCSPBypass(targetUrl) {
    try {
        // Fetch the target website
        const response = await fetch(targetUrl);
        const headers = response.headers;

        // Check if CSP header exists
        if (headers.has('content-security-policy')) {
            const cspHeader = headers.get('content-security-policy');

            // Check if 'unsafe-inline' is present in CSP header
            if (cspHeader.includes('unsafe-inline')) {
                console.log(colors.green("The website's CSP is vulnerable to 'unsafe-inline' bypass."));
                console.log("Here's how to test the vulnerability:");
                console.log(colors.yellow("1. Inspect the website's source code for inline scripts or styles."));
                console.log(colors.yellow("2. Inject a malicious script or style using the 'unsafe-inline' directive to see if it executes."));
            } else {
                console.log(colors.red("The website's CSP is not vulnerable to 'unsafe-inline' bypass."));
            }
        } else {
            console.log(colors.red("The website does not have a Content Security Policy (CSP) set."));
        }
    } catch (error) {
        console.error(colors.red("An error occurred while fetching the website:"), error);
    }
}

// Main function
async function main() {
    console.log(colors.cyan("Welcome to the CSP Bypass Vulnerability Tester!"));

    // Get target website URL from user input
    console.log(colors.cyan("Enter your target website URL: "));
    process.stdin.setEncoding('utf8');
    process.stdin.on('data', async function(data) {
        const targetUrl = data.trim();

        if (!targetUrl) {
            console.error(colors.red("Please provide a valid URL."));
            process.exit(1);
        }

        console.log(colors.yellow("Testing for CSP bypass vulnerability..."));
        await testCSPBypass(targetUrl);

        process.exit(0);
    });
}

// Execute main function
main();
